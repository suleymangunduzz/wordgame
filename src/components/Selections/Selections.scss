.Selections {
  display: flex;
  flex-flow: wrap;
  align-items: center;
  justify-content: center;
  margin-top: 1rem;

  &__value-container {
    display: flex;
    align-items: center;
  }

  &__value {
    padding: 0.3rem;
    border: 1px solid;
    margin: 0.5rem 0 0 0.3rem;
    border-radius: 0.2rem;
  }
}
