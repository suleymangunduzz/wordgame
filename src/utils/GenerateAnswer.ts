const generateAnswer = (selectedNames: string[], names: string[]) => {
  const lastValidAnswer = selectedNames[selectedNames.length - 1];

  const lastValidAnswerEndsWith = lastValidAnswer.charAt(lastValidAnswer.length - 1);

  const unusedNameList = names.filter(name => !selectedNames.includes(name));

  const newAnswer = unusedNameList.find(name => name.charAt(0) === lastValidAnswerEndsWith);

  return newAnswer || '';
};

export default generateAnswer;
