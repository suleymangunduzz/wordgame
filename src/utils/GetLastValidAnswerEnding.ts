const getLastValidAnswerEnding = (selectedNames: string[]) => {
  const lastValidAnswer = selectedNames[selectedNames.length - 1];

  return lastValidAnswer.charAt(lastValidAnswer.length - 1);
};

export default getLastValidAnswerEnding;
