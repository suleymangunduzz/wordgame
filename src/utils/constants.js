export const FINISHED = 'FINISHED';
export const NOT_STARTED = 'NOT_STARTED';
export const STARTED = 'STARTED';

export const USER = 'USER';
export const MACHINE = 'MACHINE';
