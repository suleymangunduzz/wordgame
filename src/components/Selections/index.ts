import Selections from './Selections';

export default Selections;
