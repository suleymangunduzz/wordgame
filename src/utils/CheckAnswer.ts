import getLastValidAnswerEnding from './GetLastValidAnswerEnding';
import { MACHINE, FINISHED } from './constants';

export const checkAnswer = (
  answer: string,
  selectedNames: string[],
  names: string[],
  setGameStatus: (gameStatus: string) => void,
  setNextPlayer: (nextPlayer: string) => void,
  setWinner: (winner: string) => void,
  setSelectedNames: (selectedNames: string[]) => void,
  setTimeLeft: (timeLeft: number) => any,
) => {
  const cleanAnswer = answer.replace(/\s+/g, '');
  const answerStartsWith = cleanAnswer.charAt(0);

  const lastValidAnswerEndsWith = getLastValidAnswerEnding(selectedNames);

  if (selectedNames.some(name => name === cleanAnswer)) {
    console.log('The answer is used before, so the game is finishing.');
    setGameStatus(FINISHED);
    setNextPlayer(MACHINE);
    setWinner(MACHINE);
    return;
  }

  if (answerStartsWith !== lastValidAnswerEndsWith) {
    console.log('The answer does not match with the last valid answer.');
    setGameStatus(FINISHED);
    setNextPlayer(MACHINE);
    setWinner(MACHINE);
    return;
  }

  if (!names.includes(cleanAnswer)) {
    console.log('There is no name like this in our data.');
    setGameStatus(FINISHED);
    setNextPlayer(MACHINE);
    setWinner(MACHINE);
    return;
  }

  console.log('The answer is correct and added to the our list, good job, keep playing!');
  const newState: string[] = [...selectedNames, cleanAnswer];
  setSelectedNames(newState);
  setNextPlayer(MACHINE);
  setTimeLeft(8);
};

export default checkAnswer;
