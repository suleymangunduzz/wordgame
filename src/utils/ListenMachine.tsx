import generateAnswer from './GenerateAnswer';
import { USER, FINISHED } from './constants';

export const listenMachine = (
  isLucky: boolean,
  timeToWait: number,
  selectedNames: string[],
  names: string[],
  synthRef: any,
  setWinner: (name: string) => void,
  setTimeLeft: (timeLeft: number) => void,
  setGameStatus: (status: string) => void,
  setNextPlayer: (nextPlayer: string) => void,
  setUsersFinalAnswer: (answer: string) => void,
  setSelectedNames: (selectedNames: string[]) => void,
  isFirtTime?: boolean,
) => {
  console.log('Machine is playing...');
  setUsersFinalAnswer('');

  if (isLucky) {
    console.log('Luckily, it keeps playing...');

    console.log(`After ${timeToWait} seconds machine is gonna answer.`);

    let newSelection = '';
    let newState: string[] = [];

    if (isFirtTime) {
      newSelection = names[timeToWait];
      newState = [newSelection];
    } else {
      console.log('Generate some valid name for the machine if it is possible.');
      newSelection = generateAnswer(selectedNames, names);
      newState = [...selectedNames, newSelection];
    }

    let utter: any;

    if ('SpeechSynthesisUtterance' in window) {
      utter = new SpeechSynthesisUtterance(newSelection);
      utter.lang = 'tr-TR';
    }

    setTimeout(() => {
      if (newSelection !== '') {
        setSelectedNames(newState);
        setNextPlayer(USER);
        if (synthRef.current) {
          synthRef.current.speak(utter);
        }
        setTimeLeft(8);
      } else {
        setGameStatus(FINISHED);
        setWinner(USER);
      }
    }, timeToWait * 1000);

    clearTimeout();
  } else {
    console.log('Machine can not find a name.');
    console.log('Game is gonna finish, winner is the user.');

    setTimeout(() => {
      setGameStatus(FINISHED);
      setWinner(USER);
    }, 8000);
  }
};

export default listenMachine;
