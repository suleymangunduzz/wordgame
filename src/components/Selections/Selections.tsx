import React, { FC } from 'react';

import './Selections.scss';
import { FINISHED } from '../../utils/constants';

type Props = {
  selectedNames: string[];
  gameStatus: string;
};

const Selections: FC<Props> = ({ selectedNames, gameStatus }) => {
  if (gameStatus !== FINISHED) {
    return <></>;
  }

  if (!selectedNames.length) {
    return <div className="Selections">Hiç seçim yapılamadı...</div>;
  }

  return (
    <div className="Selections">
      {selectedNames.map((name, index) => (
        <div key={name} className="Selections__value-container">
          <div className="Selections__value">{name}</div>
          {index === selectedNames.length - 1 ? '' : ' -> '}
        </div>
      ))}
    </div>
  );
};

export default Selections;
