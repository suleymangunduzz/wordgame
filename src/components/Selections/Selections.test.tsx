import React from 'react';
import { shallow } from 'enzyme';
import { NOT_STARTED, FINISHED, STARTED } from '../../utils/constants';
import Selections from './Selections';

it('renders correctly Selections when game is finished without any selection.', () => {
  const mockState: string[] = [];
  const wrapper = shallow(<Selections gameStatus={FINISHED} selectedNames={mockState} />);

  expect(wrapper).toMatchSnapshot();
});

it('renders correctly Selections when game is finished with some selections.', () => {
  const mockState: string[] = ['ayhan', 'nilay', 'yÄ±lmaz'];
  const wrapper = shallow(<Selections gameStatus={FINISHED} selectedNames={mockState} />);

  expect(wrapper).toMatchSnapshot();
});

it('renders correctly Selections when game is not started.', () => {
  const mockState: string[] = [];
  const wrapper = shallow(<Selections gameStatus={NOT_STARTED} selectedNames={mockState} />);

  expect(wrapper).toMatchSnapshot();
});

it('renders correctly Selections when game is started.', () => {
  const mockState: string[] = ['ali'];
  const wrapper = shallow(<Selections gameStatus={STARTED} selectedNames={mockState} />);

  expect(wrapper).toMatchSnapshot();
});
