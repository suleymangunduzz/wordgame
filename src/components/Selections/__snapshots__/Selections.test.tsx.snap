// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly Selections when game is finished with some selections. 1`] = `
<div
  className="Selections"
>
  <div
    className="Selections__value-container"
    key="ayhan"
  >
    <div
      className="Selections__value"
    >
      ayhan
    </div>
     -&gt; 
  </div>
  <div
    className="Selections__value-container"
    key="nilay"
  >
    <div
      className="Selections__value"
    >
      nilay
    </div>
     -&gt; 
  </div>
  <div
    className="Selections__value-container"
    key="yılmaz"
  >
    <div
      className="Selections__value"
    >
      yılmaz
    </div>
  </div>
</div>
`;

exports[`renders correctly Selections when game is finished without any selection. 1`] = `
<div
  className="Selections"
>
  Hiç seçim yapılamadı...
</div>
`;

exports[`renders correctly Selections when game is not started. 1`] = `<Fragment />`;

exports[`renders correctly Selections when game is started. 1`] = `<Fragment />`;
